add additional interpolation modes for torch.quantile

RockingJavaBean opened the pull request:
Fix #48523
Related  #38349

**BC-breaking Note:**

This PR updates PyTorch's quantile function to add additional interpolation methods `lower`, `higher`, `nearest`, and `midpoint`, and these interpolation methods are currently supported by NumPy.

New parameter `interpolation` is added to the signature for both `torch.quantile` and `torch.nanquantile` functions.

- `quantile(input, q, dim=None, interpolation='linear', keepdim=False, *, out=None) -> Tensor`
- `nanquantile(input, q, dim=None, interpolation='linear', keepdim=False, *, out=None) -> Tensor`

Function signatures followed the NumPy-like style for the moment, keeping `out` at the end to be consistent with PyTorch.

RockingJavaBean committed:
add lower, higher, midpoint, nearest interpolation for torch.quantile

RockingJavaBean committed:
add docs

RockingJavaBean commented:
The screenshots of rendered docs for the `torch.quantile` function are below.
- [description](https://user-images.githubusercontent.com/40605299/101020255-2adffe80-35a9-11eb-8aea-c9022bb9b809.png)
- [example](https://user-images.githubusercontent.com/40605299/101020515-94600d00-35a9-11eb-984f-06b347853544.png)

RockingJavaBean commented:
@mruberry  This PR implements additional quantile modes: `lower`, `higher`, `nearest`, and `midpoint` according to the function request.
Please kindly help review it.

mruberry commented:
> @mruberry This PR implements additional quantile modes: `lower`, `higher`, `nearest`, and `midpoint` according to the function request.

Awesome!

> Please kindly help review it.

@heitorschueroff, who implemented PyTorch's quantile function, will help with the review.

heitorschueroff commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -413,8 +435,9 @@ Tensor& quantile_out(
     const Tensor& self,
     const Tensor& q,
     optional<int64_t> _dim,
+    const std::string interpolation,
```
This will break backwards compatibility for the C++ API. @mruberry Is there a better solution to add this parameter than adding 8 more overloads to native_functions?

mruberry replied:
Good question. Probably not, unfortunately. Just mark the PR as "bc-breaking" with a label and add a "BC-breaking summary" at the top of the PR summary describing what's changed and why it's BC-breaking.

RockingJavaBean replied:
I updated the PR summary with bc-breaking notes for adding interpolation parameter to torch.quantile function, please kindly confirm it and may update the content.

heitorschueroff commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6728,27 +6728,27 @@
   variants: method, function
 
-- func: quantile.scalar_out(Tensor self, float q, int? dim=None, bool keepdim=False, *, Tensor(a!) out) -> Tensor(a!)
+- func: quantile.scalar_out(Tensor self, float q, int? dim=None, str interpolation='linear', bool keepdim=False, *, Tensor(a!) out) -> Tensor(a!)
```
@mruberry Do we have a preference for whether dim and keepdim should be next to each other?

mruberry replied:
It would make conceptual sense if they were next to each other. Let's implement a NumPy-like signature for the moment (keeping out at the end to be consistent with PyTorch). In the future I think we should consider making these args kwarg-only.

RockingJavaBean replied:
The NumPy's signature of quantile function is below:
`numpy.quantile(a, q, axis=None, out=None, overwrite_input=False, interpolation='linear', keepdims=False)`.

The current PR follows the NumPy style, by putting `interpolation` in front of `keepdims`.

I am a little bit uncertain about whether we should make `dim` and `keepdim` next to each other for the moment
Please kindly let me know if the order of `interpolation` and `keepdims` parameters should be exchanged, I will update this PR accordingly. @mruberry @heitorschueroff

heitorschueroff replied:
Let's keep it this way then for the moment.

heitorschueroff commented on torch/_torch_docs.py:
```diff
@@ -5116,13 +5116,20 @@ def merge_dicts(*dicts):
 Returns the q-th quantiles of each row of the :attr:`input` tensor along the dimension
-:attr:`dim`, doing a linear interpolation when the q-th quantile lies between two
-data points. By default, :attr:`dim` is ``None`` resulting in the :attr:`input` tensor
+:attr:`dim` based on :attr:`interpolation`. By default, :attr:`dim` is ``None`` resulting in the :attr:`input` tensor
```
... based on ... -> . When the desired quantile lies between two data points \`\`i < j``, the result is computed based on the :attr:\`interpolation\` value as described below. By default, :attr:\`interpolation\` is \`"linear"\` and :attr:\`dim\` ...

RockingJavaBean replied:
Addressed.

heitorschueroff commented on torch/_torch_docs.py:
```diff
@@ -5116,13 +5116,20 @@ def merge_dicts(*dicts):
+:attr:`dim` based on :attr:`interpolation`. By default, :attr:`dim` is ``None`` resulting in the :attr:`input` tensor
 being flattened before computation.
 
+The behaviors of :attr:`interpolation` are as follow:
```
Describe this in terms of data points i and j similar to how NumPy does it. Something like:

When the quantile value lies between two data points \`\`i < j\`\`, the result is computed according to the given :attr:\`interpolation\` method as follows:

Then describe each method in terms of i and j.

RockingJavaBean replied:
Thanks so much for the tip for improving the docs of `torch.quantile` function, I have pdated the docs based on the tip.

heitorschueroff commented on torch/_torch_docs.py:
```diff
@@ -5133,6 +5140,8 @@ def merge_dicts(*dicts):
     q (float or Tensor): a scalar or 1D tensor of quantile values in the range [0, 1]
     {dim}
+    interpolation (string): interpolation method to use when the desired quantile lies between two data points,
+      can be ``linear``, ``lower``, ``higher``, ``midpoint`` and ``nearest``.
```
Add "Default is ``linear``" at the end.

RockingJavaBean replied:
Addressed.

heitorschueroff commented on torch/_torch_docs.py:
```diff
@@ -5156,11 +5165,22 @@ def merge_dicts(*dicts):
     torch.Size([3, 2, 1])
+    >>> a = torch.arange(6.)
+    >>> a
+    tensor([0., 1., 2., 3., 4., 5.])
```
Great example. I would add the linear interpolation method and change the q value to 0.6 so there is a difference between linear and midpoint.

RockingJavaBean replied:
Thank you so much for pointing this out.
The linear interpolation method is added to the example, and the `q` value is changed to 0.6 to demonstrate the difference between `linear` and `midpoint`.

heitorschueroff reviewed (changes requested):
Thank you @RockingJavaBean for the PR. It's heading on the right direction but there's some room for improving performance. I also left some comments on the docs and I need to check with @mruberry on the BC breaking change.

RockingJavaBean committed:
review feedbacks

RockingJavaBean commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -91,11 +91,29 @@ void quick_select_template(
+  } else if (interpolation == "nearest") {
+    ranks.round_();
+  } else {
+    TORCH_CHECK(false, "quantile() interpolation should only be 'linear', 'lower' 'higher', "
```
Thanks for pointing this out, it has been moved next to the other checks.

RockingJavaBean commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -186,6 +204,10 @@ void quantile_impl(
   }
+
+  // adjust ranks based on the interpolation mode
+  quantile_ranks_for_interpolation(ranks, interpolation);
```
Thank you so much for providing valuable insights regarding performance optimization, and I have updated the code accordingly.

RockingJavaBean commented:
@mruberry @heitorschueroff I'm truly grateful for the kind help in reviewing this PR.
I have modified the code implementation based on the suggestions of performance optimization, the docs descriptions and examples are updated as well.
The `clang-tidy` error seems were caused by the failure of pulling the third-party code repository, and it seems is not related to this pull request.
Please kindly help review the updated PR.

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into quantile_interpolation

RockingJavaBean commented:
@heitorschueroff Gentle ping.
I have updated this PR according to the feedbacks of the code review and rebased the code, All CI checks seem passed.
Please kindly take a look.

heitorschueroff reviewed (approved):
LGTM. Thank you for this great contribution.

matthewfeickert commented:
Thanks very much for this PR. This is great to have in!

RockingJavaBean commented:
@heitorschueroff could you please share the cause of reverting this PR, is there anything I could do for help?

heitorschueroff commented:
> @heitorschueroff could you please share the cause of reverting this PR, is there anything I could do for help?

@RockingJavaBean I reverted it because some internal tests were failing and it seems this PR was to blame. I'm investigating what could be the cause and I'll let you know when I have more information.

heitorschueroff commented:
@RockingJavaBean This PR actually breaks BC compatibility of the python API as well. Before this PR, one could call `torch.quantile(samples, quantiles, 0, False)` but now this will not run because of the missing interpolation parameter. One solution is to move interpolation after keepdim but this would not be very standard with our API, @mruberry what do you think?

mruberry commented:
I think we can support current users (with a deprecation warning) and the new API by adding signatures. It's not the prettiest solution but I think we'll have to do it. It's too bad the bc-break didn't appear in any of our pre-land tests. I'm surprised it happened, but I guess that shows how popular quantile is.

heitorschueroff commented:
@RockingJavaBean I created a new PR here https://github.com/pytorch/pytorch/pull/49267 to fix the BC breaking issue. This was my fault for missing it in the review. At this point you don't need to take any further action, I'll land the PR once approved. Thanks for the contribution.

RockingJavaBean commented:
@heitorschueroff Thank you so much for all the kind help and the insightful suggestions regarding improving performances, I'm truly thankful.
